lib/oneboxer.rb

```diff
@@ -313,27 +313,82 @@ def self.allowed_iframe_origins
   def self.external_onebox(url)
     Discourse.cache.fetch(onebox_cache_key(url), expires_in: 1.day) do
       fd = FinalDestination.new(url,
-                              ignore_redirects: ignore_redirects,
-                              ignore_hostnames: blocked_domains,
-                              force_get_hosts: force_get_hosts,
-                              force_custom_user_agent_hosts: force_custom_user_agent_hosts,
-                              preserve_fragment_url_hosts: preserve_fragment_url_hosts)
+                                ignore_redirects: ignore_redirects,
+                                ignore_hostnames: blocked_domains,
+                                force_get_hosts: force_get_hosts,
+                                force_custom_user_agent_hosts: force_custom_user_agent_hosts,
+                                preserve_fragment_url_hosts: preserve_fragment_url_hosts)
       uri = fd.resolve
+
+      if fd.status != :resolved
+        args = { link: url }
+        if fd.status == :invalid_address
+          args[:error_message] = I18n.t("errors.onebox.invalid_address", hostname: fd.hostname)
+        elsif fd.status_code
+          args[:error_message] = I18n.t("errors.onebox.error_response", status_code: fd.status_code)
+        end
+
+        error_box = blank_onebox
+        error_box[:preview] = preview_error_onebox(args)
+        return error_box
+      end
+
       return blank_onebox if uri.blank? || blocked_domains.map { |hostname| uri.hostname.match?(hostname) }.any?
 
       options = {
         max_width: 695,
         sanitize_config: Onebox::DiscourseOneboxSanitizeConfig::Config::DISCOURSE_ONEBOX,
         allowed_iframe_origins: allowed_iframe_origins,
         hostname: GlobalSetting.hostname,
+        facebook_app_access_token: SiteSetting.facebook_app_access_token,
       }
 
       options[:cookie] = fd.cookie if fd.cookie
 
       r = Onebox.preview(uri.to_s, options)
+      result = { onebox: r.to_s, preview: r&.placeholder_html.to_s }
+
+      # NOTE: Call r.errors after calling placeholder_html
+      if r.errors.any?
+        # TODO: verify this will work for all locales
```

Comment:
I think this does not. But I always like to defer to @gschlager to confirm how to do this?

Reply:
Joining a sentence like that to generate a readable sentence isn't a good idea. It works for English, but might be hard to translate in other languages. Moreover, if you really wanted to do it that way, the spaces around "or" should be part of the translation.

Anyway, I suggest you rephrase the error message, so that you can append the keys at the end. Something like this: "blah blah blah error: %{keys}"

I actually like the `.to_sentence` method, but I guess a simple `.join` would work here too. It might be a good idea to add a more generic translation key to the `server.en.yml` locale file. I'd also include an example as a comment so that translators know what's going on. Something like this:

```yaml
en:
  word_connector:
    # Connects words with "or". Example: "foo or bar"
    or: " or "
```

Reply:
In the interests of (hopefully) keeping things simple, I've changed the translation to use a simple join. So the end result will be something like:

> Sorry, we were unable to generate a preview for this web page, because the following oEmbed / OpenGraph tag(s) could not be found: description

or:

> Sorry, we were unable to generate a preview for this web page, because the following oEmbed / OpenGraph tag(s) could not be found: description, image

I did consider having the 'tag(s)' as a separate attribute so that it could be pluralized before being passed in to the larger error message, but perhaps that would introduce other potential translation issues? I noticed that other error messages in Discourse use the '(s)' as a way of optionally pluralizing things (e.g. "You are not allowed to invite users to specified group(s)"), so I figure this is an acceptable compromise.